activerecord/test/cases/cache_key_test.rb

```diff
@@ -44,10 +44,88 @@ class CacheMeWithVersion < ActiveRecord::Base
 
     test "cache_key_with_version always has both key and version" do
       r1 = CacheMeWithVersion.create
-      assert_equal "active_record/cache_key_test/cache_me_with_versions/#{r1.id}-#{r1.updated_at.to_s(:usec)}", r1.cache_key_with_version
+      assert_equal "active_record/cache_key_test/cache_me_with_versions/#{r1.id}-#{r1.updated_at.utc.to_s(:usec)}", r1.cache_key_with_version
 
       r2 = CacheMe.create
-      assert_equal "active_record/cache_key_test/cache_mes/#{r2.id}-#{r2.updated_at.to_s(:usec)}", r2.cache_key_with_version
+      assert_equal "active_record/cache_key_test/cache_mes/#{r2.id}-#{r2.updated_at.utc.to_s(:usec)}", r2.cache_key_with_version
+    end
+
+    test "cache_version is the same when it comes from the DB or from the user" do
+      skip("Mysql2 does not return a string value for updated_at") if current_adapter?(:Mysql2Adapter)
+
+      record = CacheMeWithVersion.create
+      record_from_db = CacheMeWithVersion.find(record.id)
+      assert_not_called(record_from_db, :updated_at) do
+        record_from_db.cache_version
+      end
+
+      assert_equal record.cache_version, record_from_db.cache_version
+    end
+
+    test "cache_version does not truncate zeros when timestamp ends in zeros" do
+      skip("Mysql2 does not return a string value for updated_at") if current_adapter?(:Mysql2Adapter)
+
+      travel_to Time.now.beginning_of_day do
+        record = CacheMeWithVersion.create
+        record_from_db = CacheMeWithVersion.find(record.id)
+        assert_not_called(record_from_db, :updated_at) do
+          record_from_db.cache_version
+        end
+
+        assert_equal record.cache_version, record_from_db.cache_version
+      end
+    end
+
+    test "cache_version calls updated_at when the value is generated at create time" do
+      record = CacheMeWithVersion.create
+      assert_called(record, :updated_at) do
+        record.cache_version
+      end
+    end
+
+    test "cache_version does NOT call updated_at when value is from the database" do
+      skip("Mysql2 does not return a string value for updated_at") if current_adapter?(:Mysql2Adapter)
+
+      record = CacheMeWithVersion.create
+      record_from_db = CacheMeWithVersion.find(record.id)
+      assert_not_called(record_from_db, :updated_at) do
+        record_from_db.cache_version
+      end
+    end
+
+    test "cache_version does call updated_at when it is assigned via a Time object" do
+      record = CacheMeWithVersion.create
+      record_from_db = CacheMeWithVersion.find(record.id)
```

Comment:
Do these records need to be saved to the db. I think we should be able to check these behaviours on new records without saving and reloading them.

Reply:
In the `CacheMeWithVersion.create` the object assigned to `updated_at` is a Time object. That triggers the `timestamp.is_a?(String)` check which means we skip the "fast" cache behavior so then `updated_at` is guaranteed to be called. 

There might be another way to trigger this behavior that I've not found. This testing logic seemed the closest to the actual behavior I was trying to test (values generated in user space skip my logic, values from the DB use the new logic).

If you want to play around with it, the fastest way to run this test file is running this inside the AR directory:

```
bundle exec rake test:sqlite3_mem  TEST=test/cases/cache_key_test.rb
```

Reply:
@schneems For the tests that are ready values from the database I think that saving the records is required - but for the tests where values are being set by the user (as a String, Time or Hash) the tests could would with unsaved records. See https://github.com/lsylvester/rails/commit/c4ff214c741bda3e5d9957edc4bba2bb7d7c49a8

Reply:
The reason I didn't want to go with something like:

```ruby
record = CacheMeWithVersion.new
assert_called(record, :updated_at) do
  record.updated_at = Time.now.to_s
  record.cache_version
end
```

Is if the line we're testing gets removed:

```
record = CacheMeWithVersion.new
assert_called(record, :updated_at) do
  # record.updated_at = Time.now.to_s
  record.cache_version
end
```

Then the test would still pass, since the value is coming from the user already, as opposed to being from the DB. Essentially without the value being from the db it's possible to make a mistake in modifying test or in modifying the code and have all the tests still pass.